API to autocreate subnets for external hosts

rajatchopra opened the pull request:
One can now allocate hostsubnets for hosts that are not part of the cluster. This is useful when a host wants to be part of the SDN, but not part of the cluster (e.g. F5)
Trello: https://trello.com/c/fnW1tPCY/155-8-f5-integration-get-lease-from-cluster-cidr

Implementation details:
1. Allow creation of a hostsubnet resource without the 'subnet' field being present. But only if a specific annotation exists ('hostsubnet.sdn.openshift.io/autocreate').
2. Let the master also watch hostsubnets, but only bother with the ones that have the annotation. The rest of them are also created by master, but by watching nodes.
3. When the master learns about a hostsubnet being created with the annotation, it allocates a subnet of the clusterNetwork IPAM, deletes the annotated hostsubnet and recreates a new one with the assigned subnet.

The deletion and recreation have been done so that migration of existing infrastructure can go on smoothly. Even if one were to update master and not the nodes, and add an F5 machine in the middle, no issues should appear.

rajatchopra commented:
@pravisankar Thanks for the help. Please take a look.
cc @openshift/networking 

PS: This is for 3.4 release. So do not merge before the 1.3 has been cut.

liggitt commented on pkg/sdn/api/validation/validation.go:
```diff
@@ -87,7 +87,10 @@ func ValidateHostSubnet(hs *sdnapi.HostSubnet) field.ErrorList {
 	if err != nil {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("subnet"), hs.Subnet, err.Error()))
+		// check if annotation exists, then let the Subnet field be empty
+		if _, ok := hs.Annotations["hostsubnet.sdn.openshift.io/autocreate"]; !ok {
```
Why not make this an actual field?

pravisankar replied:
This annotation is mainly to address F5 use case. We use actual fields for generic use cases?

pravisankar replied:
When this annotation/field is set, ensure Subnet is empty?

rajatchopra replied:
I am in favour of an actual field. Will be easy to identify such hostsubnets easily post creation too. Thanks.

liggitt replied:
> We use actual fields for generic use cases?

at the point where API-level validation is using the data, it should probably be a field rather than an annotation

rajatchopra replied:
Then it makes me against the field idea here. This is not a useful field at runtime, its merely a direction at create time.

pravisankar commented on pkg/sdn/plugin/subnets.go:
```diff
@@ -41,6 +41,7 @@ func (master *OsdnMaster) SubnetStartMaster(clusterNetwork *net.IPNet, hostSubne
 	}
 
 	go utilwait.Forever(master.watchNodes, 0)
+	go utilwait.Forever(master.watchSubnetsOnMaster, 0)
```
go utilwait.Forever(master.watchSubnets, 0)  
OnMaster is redundant

pravisankar commented on pkg/sdn/plugin/subnets.go:
```diff
@@ -205,6 +206,47 @@ func (node *OsdnNode) initSelfSubnet() error {
+		log.V(5).Infof("Watch %s event for HostSubnet %q", strings.Title(string(eventType)), name)
+		switch eventType {
+		case watch.Added, watch.Modified:
+			if err = master.registry.ValidateNodeIP(hostIP); err != nil {
```
Not needed, addNode() validates node IP before creating subnet.

dcbw replied:
In the annotated case though, won't the subnet be created by the administrator instead, and thus the HostIP could be bogus?

pravisankar commented on pkg/sdn/plugin/subnets.go:
```diff
@@ -205,6 +206,47 @@ func (node *OsdnNode) initSelfSubnet() error {
+			if _, ok := hs.Annotations["hostsubnet.sdn.openshift.io/autocreate"]; ok {
+				err = master.registry.DeleteSubnet(name)
+				if err != nil {
+					log.Errorf("Error in recreating annotated subnet from master, name: %s, ip %s: %v", name, hostIP, err)
```
Error in deleting subnet?

pravisankar commented on pkg/sdn/plugin/subnets.go:
```diff
@@ -205,6 +206,47 @@ func (node *OsdnNode) initSelfSubnet() error {
+					log.Errorf("Error in recreating annotated subnet from master, name: %s, ip %s: %v", name, hostIP, err)
+					continue
+				}
+				err = master.addNode(name, hostIP)
```
Now that addNode() is called from watchNodes() and watchSubnets() asynchronously, we need a mutex to synchronize GetNetwork/ReleaseNetwork.

pravisankar commented on pkg/sdn/plugin/subnets.go:
```diff
@@ -205,6 +206,47 @@ func (node *OsdnNode) initSelfSubnet() error {
+				continue
+			}
+
+			if _, ok := hs.Annotations["hostsubnet.sdn.openshift.io/autocreate"]; ok {
```
Annotation used in couple of places, create a const.
We have been using "pod.network.openshift.io/" annotation prefix in other places. how about pod.network.openshift.io/autocreate-subnet?

dcbw replied:
Yeah, needs a constant, and we should be consistent with other constants like the macvlan one.

pravisankar commented:
In SubnetStartMaster(), you need to ignore HostSubnets that doesn't have populated Subnet field.

pravisankar commented on pkg/sdn/plugin/subnets.go:
```diff
@@ -205,6 +206,47 @@ func (node *OsdnNode) initSelfSubnet() error {
+			}
+
+			if _, ok := hs.Annotations["hostsubnet.sdn.openshift.io/autocreate"]; ok {
+				err = master.registry.DeleteSubnet(name)
```
Add a comment on why we are doing delete+create subnet instead of update subnet.

danwinship commented:
> 1. Allow creation of a hostsubnet resource without the 'subnet' field being present. But only if a specific annotation exists ('hostsubnet.sdn.openshift.io/autocreate').

Why not just drop the annotation, and say people with HostSubnet-creating permission can always create HostSubnets-without-Subnet-fields, and the master will always then fill in the missing field?

> deletes the annotated hostsubnet and recreates a new one with the assigned subnet.

Why does it do this rather than just update the existing one?

rajatchopra commented:
Fixed the issues in feedback.
@danwinship We could drop the annotation and be permissive in allowing that hostsubnets be created without subnet being defined. The annotation method just makes it more explicit with respect to the intent. It will prevent the accidental case of subnet gone missing in another case.

Also, added a comment about why a delete+create is done instead of an update. The only reason is to avoid migration issues. The node side watchSubnets skips the actual planting of OVS rules if the hostsubnet was modified but the host name remains the same.

pravisankar commented:
LGTM, update docs about this feature?

rajatchopra commented:
Yes, docs need to be updated. But, kind of a half story right now without the F5 use case. The complete F5 native-vxlan thing is coming in another PR, so probably a doc when that is ready. Thanks for the feedback.

rajatchopra commented:
[test]

rajatchopra commented:
re [test]

liggitt commented on pkg/sdn/api/validation/validation.go:
```diff
@@ -87,7 +88,10 @@ func ValidateHostSubnet(hs *sdnapi.HostSubnet) field.ErrorList {
 	if err != nil {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("subnet"), hs.Subnet, err.Error()))
+		// check if annotation exists, then let the Subnet field be empty
+		if _, ok := hs.Annotations[sdnplugin.AssignHostSubnetAnnotation]; !ok {
```
this isn't just allowing the subnet field to be empty, it's allowing it to be invalid... that's not what we want is it?

rajatchopra commented:
Refactored with feedback and the restructure from eventQueue to DeltaFifo

rajatchopra commented:
[test]

knobunc commented:
[test]

danwinship commented on pkg/sdn/plugin/subnets.go:
```diff
@@ -240,6 +241,41 @@ func (node *OsdnNode) initSelfSubnet() error {
 }
 
+// Only run on the master
+// Watch for all hostsubnet events and if one is found with the right annotation, use the IPAM to dole a real subnet
```
"use the subnet allocator" not "use the IPAM". "IPAM" refers to assigning IPs to pods, not networks to nodes

knobunc commented:
[test]

rajatchopra commented:
re [test]

rajatchopra commented:
re [test]
Flake:  https://github.com/openshift/origin/issues/9548 https://github.com/openshift/origin/issues/9490

knobunc commented:
re [test]

knobunc commented:
Flake https://github.com/openshift/origin/issues/11315 [merge]
